presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java

```diff
@@ -181,7 +182,15 @@ public ActualProperties visitEnforceSingleRow(EnforceSingleRowNode node, List<Ac
         @Override
         public ActualProperties visitAssignUniqueId(AssignUniqueId node, List<ActualProperties> inputProperties)
         {
-            return Iterables.getOnlyElement(inputProperties);
+            ActualProperties properties = Iterables.getOnlyElement(inputProperties);
+            if (properties.getNodePartitioning().isPresent()) {
+                // preserve input (possibly preferred) partitioning
+                return properties;
+            }
+
+            return ActualProperties.builderFrom(properties)
+                    .global(partitionedOn(FIXED_HASH_DISTRIBUTION, ImmutableList.of(node.getIdColumn()), Optional.empty()))
```

Comment:
FIXED_HASHED_DISTRIBUTION is not right. AssignUniqueId could be in a leaf stage.

Reply:
`Global#nodePartitioning` currently doesn't allow to specify an unknown partition method on a set of columns. Can we extend `Global` to allow that? @martint @sopel39

Reply:
I would just add
```
ARBITRARY_DISTRIBUTION = createSystemPartitioning(SystemPartitioning.ARBITRARY, SystemPartitionFunction.UNKNOWN);
```
 
> Global#nodePartitioning currently doesn't allow to specify an unknown partition method on a set of columns. Can we extend Global to allow that

This could be a major `ActualProperties` refactor for a relatively simple change.

Reply:
@sopel39 

> ARBITRARY_DISTRIBUTION = createSystemPartitioning(SystemPartitioning.ARBITRARY, SystemPartitionFunction.UNKNOWN);

That should work.

Reply:
> Global#nodePartitioning currently doesn't allow to specify an unknown partition method on a set of columns. Can we extend Global to allow that?

Interestingly, this used to be possible. We could express thing such as "partitioned on these columns, but don't know how" and "partitioned on these columns using a hash function that evaluates them in this specific order". The latter was a stronger form that satisfied requirements that needed the former. I'd like to figure out where we lost that capability.

Reply:
> Interestingly, this used to be possible. We could express thing such as "partitioned on these columns, but don't know how" and "partitioned on these columns using a hash function that evaluates them in this specific order". The latter was a stronger form that satisfied requirements that needed the former. I'd like to figure out where we lost that capability.

I think `Partitioning` in `ActualProperties` was introduced by `cc9fd708fff2c075a96caa87636cee431a114018` and seems quite old.

@martint let me know if `ARBITRARY_DISTRIBUTION` is ok for now.